### PR TITLE
Allow dumping TSEC keys to sd card

### PR DESCRIPTION
**Pull request by tesnos**
It would be useful if we could dump the TSEC keys just like the fuses and kfuses

### CODE REVIEW

File: ipl/main.c [hunk 1]
```diff
@@ -665,31 +665,52 @@ void print_tsec_key()
 	{
 		EPRINTFARGS("Could not identify package1 version\nto read TSEC firmware (= '%s').",
 			(char *)pkg1 + 0x10);
+		
+		gfx_puts(&gfx_con, "\nPress any key...\n");
+		btn_wait();
```
Comment (CTCaer) [inline ipl/main.c]: Remove these 2.
Press any key is not needed anymore.
btn_wait is going to be used elsewhere.

File: ipl/main.c [hunk 2]
```diff
@@ -665,31 +665,52 @@ void print_tsec_key()
...
+			for (u32 j = 0; j < 0x10; j++)
+				gfx_printf(&gfx_con, "%02X", keys[((i - 1) * 0x10) + j]);
 		}
 		else
 			EPRINTFARGS("ERROR %X", res);
 		gfx_putc(&gfx_con, '\n');
 	}
+	
+		free(pkg1);
+	sdmmc_storage_end(&storage);
```
Comment (CTCaer) [inline ipl/main.c]: Remove these 2.
They create undefined behavior when these 2 and the 2 at `out` label run.

File: ipl/main.c [hunk 3]
```diff
@@ -665,31 +665,52 @@ void print_tsec_key()
...
+			char tsec_keyFilename[26];
+			f_mkdir("Backup");
+			f_mkdir("Backup/Dumps");
+			memcpy(tsec_keyFilename, "Backup/Dumps/tsec_key.bin", 26);
+
+			if (!sd_save_to_file(keys, 0x10 * 3, tsec_keyFilename))
+				gfx_puts(&gfx_con, "\nDone!\n");
+			sd_unmount();
+		}
+	}
```
Comment (CTCaer) [inline ipl/main.c]:
```
else
	goto out;
```

File: ipl/main.c [hunk 4]
```diff
@@ -665,31 +665,52 @@ void print_tsec_key()
...
 
 out:;
```
Comment (CTCaer) [inline ipl/main.c]: Create a label `out_wait:;` above label `out`
and add `btn_wait()` below it.

File: ipl/main.c [hunk 5]
```diff
@@ -665,31 +665,52 @@ void print_tsec_key()
 	{
 		EPRINTFARGS("Could not identify package1 version\nto read TSEC firmware (= '%s').",
 			(char *)pkg1 + 0x10);
+		
+		gfx_puts(&gfx_con, "\nPress any key...\n");
+		btn_wait();
 		goto out;
```
Comment (CTCaer) [inline ipl/main.c]: change to `goto out_wait`

**Comment by tesnos** [comment]
Thank you for the comments, the requested changes have been made. Plus an additional change of moving `sdmmc_storage_end(&storage);` higher up in the code. This was done due to the payload freezing after dumping to the sd card and then running that line.  The storage pointer is used no farther than where the new location is so it is safe and it made no change in the size of the payload.  I compiled and tested it on my own switch and was able to both exit the menu without dumping and exit the menu after dumping as expected

**Comment by CTCaer** [comment]
Thank you!

(Btw, `sdmmc_storage_end(&storage);` should always work, it's strange that it hanged.)

**Comment by fennectech** [comment]
Nice :D

**Pull request merged**